File: src/linksearch/platforms/ddg_social.py
from __future__ import annotations

import logging
import time
from typing import TYPE_CHECKING

from linksearch.models import CandidateLink, ProductInput
from linksearch.platforms.google_cse import (
    build_platform_search_queries,
    url_matches_site_host,
)

if TYPE_CHECKING:
    from linksearch.config import Settings

logger = logging.getLogger(__name__)

MAX_DDG_QUERIES_PER_SITE = 4
DDG_DELAY_SECONDS = 0.35


def search_site_ddg_sync(
    settings: Settings,
    product: ProductInput,
    queries: list[str],
    site_host: str,
    platform_keyword: str,
    media_label: str,
    *,
    max_queries: int | None = None,
    result_cap: int | None = None,
) -> list[CandidateLink]:
    """
    Discover links via DuckDuckGo text search (free, no Google Custom Search API).
    Keeps URLs whose host matches ``site_host``.
    """
    if not settings.ddg_social_enabled:
        return []
    try:
        from duckduckgo_search import DDGS
    except ImportError:
        logger.warning(
            "duckduckgo-search not installed; skip DDG social search (pip install duckduckgo-search)."
        )
        return []

    mq_limit = max_queries if max_queries is not None else MAX_DDG_QUERIES_PER_SITE
    search_queries = build_platform_search_queries(platform_keyword, product, queries)[
        :mq_limit
    ]
    if not search_queries:
        return []

    cap = max(1, result_cap if result_cap is not None else settings.max_results_per_platform)
    out: list[CandidateLink] = []
    seen: set[str] = set()
    brand = product.normalized_brand()
    sku = product.sku.strip()
    pname = product.product_name.strip()

    for q in search_queries:
        if len(out) >= cap:
            break
        full_q = f"site:{site_host} {q}"[:400]
        try:
            time.sleep(DDG_DELAY_SECONDS)
            with DDGS() as ddgs:
                raw = ddgs.text(full_q, max_results=min(20, cap * 3))
                results = list(raw) if raw is not None else []
        except Exception as e:
            logger.warning("DDG search failed (%s): %s", full_q[:80], e)
            continue

        for r in results:
            if len(out) >= cap:
                break
            if not isinstance(r, dict):
                continue
            url = str(r.get("href") or r.get("url") or "").strip()
            if not url or url in seen:
                continue
            if not url_matches_site_host(url, site_host):
                continue
            seen.add(url)
            title = str(r.get("title") or "")
            body = str(r.get("body") or "")[:500]
            out.append(
                CandidateLink(
                    media=media_label,
                    brand=brand,
                    url=url,
                    sku=sku,
                    product_name=pname,
                    title=title,
                    snippet=body,
                    source_query=full_q,
                )
            )

    return out


def search_tiktok_ddg(
    settings: Settings,
    product: ProductInput,
    queries: list[str],
    *,
    max_queries: int | None = None,
    result_cap: int | None = None,
) -> list[CandidateLink]:
    return search_site_ddg_sync(
        settings,
        product,
        queries,
        "tiktok.com",
        "Tiktok",
        "Tiktok",
        max_queries=max_queries,
        result_cap=result_cap,
    )


def search_facebook_ddg(
    settings: Settings,
    product: ProductInput,
    queries: list[str],
    *,
    max_queries: int | None = None,
    result_cap: int | None = None,
) -> list[CandidateLink]:
    return search_site_ddg_sync(
        settings,
        product,
        queries,
        "facebook.com",
        "Facebook",
        "Facebook",
        max_queries=max_queries,
        result_cap=result_cap,
    )


def search_instagram_ddg(
    settings: Settings,
    product: ProductInput,
    queries: list[str],
    *,
    max_queries: int | None = None,
    result_cap: int | None = None,
) -> list[CandidateLink]:
    return search_site_ddg_sync(
        settings,
        product,
        queries,
        "instagram.com",
        "Instagram",
        "Instagram",
        max_queries=max_queries,
        result_cap=result_cap,
    )

File: frontend/src/lib/platforms/ddgSocial.ts
import { search, SafeSearchType } from "duck-duck-scrape";

import type { CandidateLink, ProductInput, SearchSettings } from "../types";
import {
  buildPlatformSearchQueries,
  urlMatchesSiteHost,
} from "./googleCse";

const MAX_DDG_QUERIES = 4;
const DDG_MS = 350;

function sleep(ms: number) {
  return new Promise((r) => setTimeout(r, ms));
}

async function searchSiteDdg(
  settings: SearchSettings,
  product: ProductInput,
  queries: string[],
  siteHost: string,
  platformKeyword: string,
  mediaLabel: string
): Promise<CandidateLink[]> {
  if (!settings.ddgSocialEnabled) return [];

  const searchQueries = buildPlatformSearchQueries(
    platformKeyword,
    product,
    queries
  ).slice(0, MAX_DDG_QUERIES);
  const cap = settings.maxResultsPerPlatform;
  const out: CandidateLink[] = [];
  const seen = new Set<string>();
  const brand = product.brand.trim();
  const sku = product.sku.trim();
  const productName = product.productName.trim();

  for (const q of searchQueries) {
    if (out.length >= cap) break;
    const fullQ = `site:${siteHost} ${q}`.slice(0, 400);
    try {
      await sleep(DDG_MS);
      const data = await search(fullQ, {
        safeSearch: SafeSearchType.MODERATE,
      });
      for (const r of data.results ?? []) {
        if (out.length >= cap) break;
        const url = String(r.url ?? "").trim();
        if (!url || seen.has(url)) continue;
        if (!urlMatchesSiteHost(url, siteHost)) continue;
        seen.add(url);
        out.push({
          media: mediaLabel,
          brand,
          url,
          sku,
          productName,
          title: String(r.title ?? ""),
          snippet: String(r.description ?? "").replace(/<\/?b>/gi, "").slice(0, 500),
          score: 0,
          sourceQuery: fullQ,
        });
      }
    } catch {
      /* DDG may rate-limit or block; continue */
    }
  }

  return out;
}

export function searchTiktokDdg(
  s: SearchSettings,
  p: ProductInput,
  q: string[]
) {
  return searchSiteDdg(s, p, q, "tiktok.com", "Tiktok", "Tiktok");
}

export function searchFacebookDdg(
  s: SearchSettings,
  p: ProductInput,
  q: string[]
) {
  return searchSiteDdg(s, p, q, "facebook.com", "Facebook", "Facebook");
}

export function searchInstagramDdg(
  s: SearchSettings,
  p: ProductInput,
  q: string[]
) {
  return searchSiteDdg(s, p, q, "instagram.com", "Instagram", "Instagram");
}

/** YouTube via DuckDuckGo `site:youtube.com` — works without API key or yt-dlp (e.g. Vercel). */
export function searchYoutubeDdg(
  s: SearchSettings,
  p: ProductInput,
  q: string[]
) {
  return searchSiteDdg(s, p, q, "youtube.com", "YouTube", "Youtube");
}

/** Shorts-shaped URLs only (`/shorts/`). Pairs with DDG when yt-dlp is unavailable. */
export async function searchYoutubeShortsDdg(
  s: SearchSettings,
  p: ProductInput,
  q: string[]
): Promise<CandidateLink[]> {
  const rows = await searchSiteDdg(
    s,
    p,
    q,
    "youtube.com",
    "YouTube Shorts",
    "YoutubeShorts"
  );
  return rows.filter((r) => /\/shorts\//i.test(r.url));
}
